worldedit-core/src/main/java/com/fastasyncworldedit/core/util/task/KeyQueuedExecutorService.java

```diff
@@ -0,0 +1,164 @@
+    }
+
+    /**
+     * Delegates to {@link ThreadPoolExecutor#isShutdown()}
+     */
+    public boolean isShutdown() {
+        return parent.isShutdown();
+    }
+
+    /**
+     * Delegates to {@link ThreadPoolExecutor#isTerminated()}
+     */
+    public boolean isTerminated() {
+        return parent.isTerminated();
+    }
+
+    /**
+     * Delegates to {@link ThreadPoolExecutor#awaitTermination(long, TimeUnit)}
+     */
+    public boolean awaitTermination(long timeout, @Nonnull TimeUnit unit) throws InterruptedException {
+        return parent.awaitTermination(timeout, unit);
+    }
+
+    protected <T> FutureTask<T> newTaskFor(Runnable runnable, T value) {
+        return new FutureTask<>(runnable, value);
+    }
+
+    protected <T> FutureTask<T> newTaskFor(Callable<T> callable) {
+        return new FutureTask<>(callable);
+    }
+
+    @Nonnull
+    public <T> Future<T> submit(@Nonnull K key, @Nonnull Callable<T> task) {
+        FutureTask<T> ftask = newTaskFor(task);
+        execute(key, ftask);
+        return ftask;
+    }
+
+    @Nonnull
+    public <T> Future<T> submit(@Nonnull K key, @Nonnull Runnable task, T result) {
+        FutureTask<T> ftask = newTaskFor(task, result);
+        execute(key, ftask);
+        return ftask;
+    }
+
+    @Nonnull
+    public Future<?> submit(@Nonnull K key, @Nonnull Runnable task) {
+        FutureTask<Void> ftask = newTaskFor(task, null);
+        execute(key, ftask);
+        return ftask;
+    }
+
+    public void execute(@Nonnull K key, @Nonnull FutureTask<?> command) {
+        KeyRunner runner;
+        synchronized (keyQueue) {
+            runner = keyQueue.computeIfAbsent(key, KeyRunner::new);
+            runner.add(command);
+            runner.triggerRun();
+        }
+    }
+
+    private final class KeyRunner {
+
+        private final Queue<FutureTask<?>> tasks = new ConcurrentLinkedQueue<>();
+        private final K key;
+
+        private KeyRunner(K key) {
+            this.key = key;
+        }
+
+        void add(FutureTask<?> task) {
+            if (!tasks.add(task)) {
+                throw new RejectedExecutionException(rejection());
+            }
+        }
+
+        void triggerRun() {
+            Runnable task = tasks.poll();
+            if (task == null) {
+                throw new RejectedExecutionException(rejection());
+            }
+            try {
+                run(task);
+            } catch (RejectedExecutionException e) {
+                synchronized (keyQueue) {
+                    keyQueue.remove(key);
+                }
+                throw new RejectedExecutionException(rejection(), e);
+            }
+        }
+
+        private void run(Runnable task) {
+            parent.execute(() -> {
+                task.run();
+                Runnable next = tasks.poll();
+                if (next == null) {
+                    synchronized (keyQueue) {
+                        next = tasks.remove();
```

Comment:
I think this is still wrong. If the queue is still empty here, it throws an exception instead of removing the key.

Reply:
Meant to replace all remove with poll, mb